src/transformations/defmodel.cpp

```diff
@@ -0,0 +1,445 @@
+            std::unique_ptr<NS_PROJ::GenericShiftGridSet> &&realGridSetIn)
+        : ctx(ctxIn), realGridSet(std::move(realGridSetIn)) {}
+
+    const Grid *gridAt(double x, double y) {
+        const NS_PROJ::GenericShiftGrid *realGrid = realGridSet->gridAt(x, y);
+        if (!realGrid) {
+            return nullptr;
+        }
+        auto iter = mapGrids.find(realGrid);
+        if (iter == mapGrids.end()) {
+            iter = mapGrids
+                       .insert(std::pair<const NS_PROJ::GenericShiftGrid *,
+                                         std::unique_ptr<Grid>>(
+                           realGrid,
+                           std::unique_ptr<Grid>(new Grid(ctx, realGrid))))
+                       .first;
+        }
+        return iter->second.get();
+    }
+
+  private:
+    GridSet(const GridSet &) = delete;
+    GridSet &operator=(const GridSet &) = delete;
+};
+
+struct EvaluatorIface : public EvaluatorIfaceConcept<Grid, GridSet> {
+
+    PJ_CONTEXT *ctx;
+    PJ *cart;
+
+    EvaluatorIface(PJ_CONTEXT *ctxIn, PJ *cartIn) : ctx(ctxIn), cart(cartIn) {}
+
+    ~EvaluatorIface() {
+        if (cart)
+            cart->destructor(cart, 0);
+    }
+
+    std::unique_ptr<GridSet> open(const std::string &filename) {
+        auto realGridSet = NS_PROJ::GenericShiftGridSet::open(ctx, filename);
+        if (!realGridSet) {
+            pj_log(ctx, PJ_LOG_ERROR, "defmodel: cannot open %s",
+                   filename.c_str());
+            return nullptr;
+        }
+        return std::unique_ptr<GridSet>(
+            new GridSet(ctx, std::move(realGridSet)));
+    }
+
+    bool isGeographicCRS(const std::string &crsDef) {
+        PJ *P = proj_create(ctx, crsDef.c_str());
+        if (P == nullptr) {
+            return true; // reasonable default value
+        }
+        const auto type = proj_get_type(P);
+        bool ret = (type == PJ_TYPE_GEOGRAPHIC_2D_CRS ||
+                    type == PJ_TYPE_GEOGRAPHIC_3D_CRS);
+        proj_destroy(P);
+        return ret;
+    }
+
+    void geographicToGeocentric(double lam, double phi, double height, double a,
+                                double b, double /*es*/, double &X, double &Y,
+                                double &Z) {
+        (void)a;
+        (void)b;
+        assert(cart->a == a);
+        assert(cart->b == b);
+        PJ_LPZ lpz;
+        lpz.lam = lam;
+        lpz.phi = phi;
+        lpz.z = height;
+        PJ_XYZ xyz = cart->fwd3d(lpz, cart);
+        X = xyz.x;
+        Y = xyz.y;
+        Z = xyz.z;
+    }
+
+    void geocentricToGeographic(double X, double Y, double Z, double a,
+                                double b, double /*es*/, double &lam,
+                                double &phi, double &height) {
+        (void)a;
+        (void)b;
+        assert(cart->a == a);
+        assert(cart->b == b);
+        PJ_XYZ xyz;
+        xyz.x = X;
+        xyz.y = Y;
+        xyz.z = Z;
+        PJ_LPZ lpz = cart->inv3d(xyz, cart);
```

Comment:
Can this fail? What's its failure condition and can we recover?

Reply:
Lucikly, forward and inverse cart can't fail !